core/src/connection_reuse.rs

```diff
@@ -281,145 +487,312 @@ where
         loop {
             match self.listener.poll() {
                 Ok(Async::Ready(Some(upgrade))) => {
-                    self.current_upgrades.push(upgrade);
+                    trace!("New incoming connection");
+                    self.current_upgrades.push(Box::new(upgrade));
                 }
                 Ok(Async::NotReady) => break,
                 Ok(Async::Ready(None)) => {
-                    debug!("listener has been closed");
-                    if self.connections.is_empty() && self.current_upgrades.is_empty() {
-                        return Ok(Async::Ready(None));
-                    }
-                    break
+                    debug!("Listener has been closed");
+                    break;
                 }
                 Err(err) => {
-                    debug!("error while polling listener: {:?}", err);
-                    if self.connections.is_empty() && self.current_upgrades.is_empty() {
-                        return Err(err);
-                    }
-                    break
+                    debug!("Error while polling listener: {:?}", err);
+                    return Err(err);
                 }
-            }
+            };
         }
 
+        // Process the connections being upgraded.
         loop {
             match self.current_upgrades.poll() {
                 Ok(Async::Ready(Some((muxer, client_addr)))) => {
+                    // Successfully upgraded a new incoming connection.
+                    trace!("New multiplexed connection from {}", client_addr);
+                    let mut shared = self.shared.lock();
                     let next_incoming = muxer.clone().inbound();
-                    self.connections
-                        .push((muxer.clone(), next_incoming, client_addr.clone()));
+                    let connection_id = shared.next_connection_id;
+                    shared.next_connection_id += 1;
```

Comment:
seeing how often this
```rust
                    let connection_id = shared.next_connection_id;
                    shared.next_connection_id += 1;
```
appears in the code, I feel, this should be bundled in a helper function direclty on `shared` rather than being used directly to prevent misuse.